File: software/examples/c/example_callback.c

#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_motion_detector.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change to your UID


// Callback function for end of detection cycle 
void cb_detection_cycle_ended(void *user_data) {
	(void)user_data; // avoid unused parameter warning

	printf("Detection Cycle Ended (next detection possible in ~3 seconds)\n");
}

// Callback function for detected motion
void cb_motion_detected(void *user_data) {
	(void)user_data; // avoid unused parameter warning

	printf("Motion Detected\n");
}

int main() {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	MotionDetector md;
	motion_detector_create(&md, UID, &ipcon); 

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		exit(1);
	}
	// Don't use device before ipcon is connected

	// Register detected callback to function cb_motion_detected
	motion_detector_register_callback(&md,
	                                  MOTION_DETECTOR_CALLBACK_MOTION_DETECTED,
	                                  (void *)cb_motion_detected,
	                                  NULL);

	// Register detection cycle ended callback to function cb_detection_cycle_ended
	motion_detector_register_callback(&md,
	                                  MOTION_DETECTOR_CALLBACK_DETECTION_CYCLE_ENDED,
	                                  (void *)cb_detection_cycle_ended,
	                                  NULL);

	printf("Press key to exit\n");
	getchar();
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
}

File: software/examples/python/example_callback.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-  

HOST = "localhost"
PORT = 4223
UID = "XYZ" # Change to your UID

from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_motion_detector import MotionDetector

# Callback function for end of detection cycle
def cb_detection_cycle_ended():
    print('Detection Cycle Ended (next detection possible in ~3 seconds)')

# Callback function for detected motion
def cb_motion_detected():
    print('Motion Detected')

if __name__ == "__main__":
    ipcon = IPConnection() # Create IP connection
    md = MotionDetector(UID, ipcon) # Create device object

    ipcon.connect(HOST, PORT) # Connect to brickd
    # Don't use device before ipcon is connected

    # Register detected callback to function cb_motion_detected
    md.register_callback(md.CALLBACK_MOTION_DETECTED, cb_motion_detected)

    # Register detection cycle ended callback to function cb_detection_cycle_ended
    md.register_callback(md.CALLBACK_DETECTION_CYCLE_ENDED, cb_detection_cycle_ended)

    raw_input('Press key to exit\n') # Use input() in Python 3
    ipcon.disconnect()
